### ISSUE TITLE
beta7 Sites now operate, but assets include site introducer, so they don't.

### ISSUE THREAD

**Issue by narration-sd** [craftcms/cms#1477]
### Description
With more than the default Site, image assets don't show on the added Site/s.
printing out their entry.assetId.one.url show that the Site subfolder is not removed for added sites, ergo reaches nothing. This is with the initial Sites fix from #1472.

### Steps to reproduce

1. Have at least a second Site defined and prepared with appropriate introducer  subfolder & contents.
2. Have an image asset field, and try to show it on a twig template
3. Print or inspecct the asset url, and see it unexpectedly includes the  introducer folder name.

### Additional info

PHP version	7.1.2-3+deb.sury.org~xenial+1
Database driver & version	MySQL 5.7.17-0ubuntu0.16.04.1
Image driver & version	GD 7.1.2-3+deb.sury.org~xenial+1
Craft edition & version	Craft Personal 3.0.0-beta.7
Yii version	2.0.11.2
Twig version	2.1.0
Guzzle version	6.2.1
Imagine version	0.7-dev
Plugins: none

**Comment by brandonkelly** [comment]
What is your asset volume’s Base URL?

Guessing this is a duplicate of #1462.

**Comment by narration-sd** [comment]
Well, I surely thought you had it, but:
- asset url def in volumes.php was already @web/resources/images
- as shown in Settings/Assets, /resources/images (?? - guess the @web might be paranthetical there to show where it got it for sure...
- I tried setting ASSETS_BASE_URL to '/', along with CRAFT_SITE in the introducer, but no joy
- a light bulb went off about  something Brad hinted me in early days, so I tried setting     'resourceTrigger' => 'myresources', in general.php - no change
- all these steps with clear caches and reindex assets after, just in case

Here's a pic of my instrumented test page output, if useful:

![casa-real-assets-nope](https://cloud.githubusercontent.com/assets/247945/23815909/b791eda0-059e-11e7-9e63-323a50a8cf95.png)

**Comment by narration-sd** [comment]
p.s. what's it doing looking at cpresources -- just  the transform def, not the resultant transformed image if transform had gone through?

**Comment by andris-sevcenko** [comment]
/cpresources/transforms/X is the url that Craft uses to start a transform with the id of X. When the transform is done, that URL would redirect to the actual transform.

If you hate those URLs, you can set `generateTransformsBeforePageLoad` to true.

**Comment by narration-sd** [comment]
Thanks, Andris - it was just a curiosity in the middle of trying to nail down the actual problem/s here.

Don't hate the links, but it's a good idea to try out the pre-generation, and how that situation works.

Take care, working on Saturday..!

**Comment by brandonkelly** [comment]
> asset url def in volumes.php was already @web/resources/images

`@web` will refer to whatever the base URL that the current `index.php` lives in – and `@webroot` refers to the server path to the same directly. So I guess it’s not a good idea to use those if you have multiple `index.php` entry points.

> I tried setting ASSETS_BASE_URL to '/', along with CRAFT_SITE in the introducer, but no joy

Craft doesn't support a `ASSETS_BASE_URL` constant; is that a custom thing? Where are you defining it, and where are you outputting its value? (Perhaps in `config/volumes.php`?)

> a light bulb went off about something Brad hinted me in early days, so I tried setting 'resourceTrigger' => 'myresources', in general.php - no change

As @andris-sevcenko said, the cpresources/ URL is what is supposed to trigger asset transform generation (if the transform doesn't exist yet).

I'm guessing there are two parts to this issue. First, if your volume's file server path uses the `@webroot` alias, then Craft isn't going to be able to find your assets on the file system in the first place, since it will be looking from the wrong starting location. Second, even if it found the asset, it would then be redirecting the browser to the wrong URL, because `@web` is pointing to the wrong directory.

So, for multi-site, avoid those two alias, and stick with hard-coding the absolute URL & path to your root volume folder in the volume settings (either in the CP or from `config/volumes.php`).

**Comment by narration-sd** [comment]
Alright you kookoobirds, yes, I agree with your close, if a bit early since not confirmed.

However, it confirms, and when this menagerie of configuration is suitable, the multiSite works very nicely, feels 'simple'. If I have a few more corners to tie down fully, but expect it will all be fine now. So there, I did 'it' too.

Here are I hope a set of short observations and questions which I hope you feel are germane:

- One thing that makes this situation particularly confuting is that even when @web prefixes the url definition in volumes.php is that the same @web doesn't show on the CP screen for the Asset bundle. I think I would remedy that, or plan to hear about it, not to mention secondhand smoke damage coming from your software clients' ears.

- Left hanging here is the question of using @webroot in parallel or not in parallel for the bundle path?? I left in just to see, with no issue so far. I expect it should come out, and that these two @web* definitions should work similarly, track???

- Ok, now we have not-in-public nailed down how the basic config should work, and the underlying beta thing that left this in knots originally. 'What about when you do, per Site, want to use (and some only especially) custom Assets? My first thought is that you then need a futher asset bundle def for it, and to switch to it in Twig based on Site, for those particular site-orientated assets? 

- Ok again, you to expect some kind of general comment, when thought is coming from this quarter. It seems to form up this time as:

 'You guys, some in particular, are very good at coming up with useful and well-targeted abstractions, generally very solidly designed and implemented. You are rather less facile at ever telling what they are, leaving it to client mumblety-peg with various configuring parts to figure something out themselves'. 

This after I woke ready to try and/or trace pretty precisely what the problem turned out to be this time, in the vicinity of @web. It does get tedious in the revolving code worlds of internally setting up Craft to run, though I have the patient energy, and if in this case we'd not be looking for something actually (and conceptually missing), as Ben and I were on the top end of this.. It is however a pretty huge time sink, however much is usefully learned each time I do it, another reason I like to do it. 

You want to remember also that there can be more layers involved even than you invent -- nginx and getting the parts Craft needs to work with (and different config to use them), one of the layers in this case. Yes, the info was out there on the web, finally, but rather unsure of itself, many bad paths taken, and generally not the sort of thing mid-level developers are going to untangle, at the same time as nginx for many good reasons is leaving apache behind.

So what am I suggesting?  I find the kind of definition I found myself drawing out out for the Sites abstraction has, as really it should, very little to do with the code/config details we got hung  up on in these issues reported. And I remind you that even you guys were tangled. 

Defining abstractions, though, is really the highest  level of the game, my view anyway, and makes for the clarity both you and your clients can use. When you do it, you have a great place to begin (and often end) some typically much briefer doc -- and have it early. Given you also write very well about what you do say, this should be a big positive all around.

Let me leave this ;'abstract' itself, then, and maybe that will help reckoning to the point. 

Software has so often gone the other way, and even or especially in early Unix days the only solution was to read the code --- where you discovered I remember Berkeley grad students writing truly atrocious hacks that only worked if you took the one path through the problem they had  actually thought out.. 

It's quite possible my perspective here comes from being more a words guy than a code guy, though I put together the two quite often in my own better achievements, would judge from distance anyway. So if this essay has been useful...as I always intend, guys, I think you know.

C.

**Comment by narration-sd** [comment]
afterthought: code reading will always be a part of working out things like the nginx; the connectivity with unanticipated peripheral systems. But again, I think you can help with that, by being more forthcoming with the _intent_ of Craft's orten-object concepts, and of the configuration items themselves.

**Comment by narration-sd** [comment]
Also, the point above about different Asset bundles per Site is not needed -- we are already able just to set the different assets in our Site-able fields. One  thing I was originally testing where this all broke down, so already in place.

But it's an illustration about  how concepts linger which may not be well-formed unless you form them: comes in this case (that was the memory mentioned) of your original enthusiastic public promo of the new Sites siystem and its capabilities.  So that's where the second level, carefully laying out what the abstraction is and does, comes -- prior to telling (and informing, clarifying  naturally) what you are going to say about what individual configs tune...
